File: ChessEngine/rook.py
rook_offsets = [(1, 0), (-1, 0), (0, 1), (0, -1)]

def check_rook(current_board: dict ,moves_list: dict, value="rook"):
    """
    The function `check_rook` iterates through a chess board to find valid moves for rooks and updates a
    moves list accordingly.
    
    :param current_board: The function `check_rook` seems to be checking for valid moves for rooks on a
    chessboard based on the provided `current_board` dictionary. It appears to be iterating over the
    board positions and checking for possible rook moves in horizontal and vertical directions
    :type current_board: dict
    :param moves_list: The `moves_list` parameter in the `check_rook` function seems to be a dictionary
    that contains lists of moves for white ('w') and black ('b') players. The function appears to be
    checking for valid rook moves on a chessboard based on the current board state provided in the
    :type moves_list: list
    :param value: The `value` parameter in the `check_rook` function is used to specify the type of
    piece being checked for on the board. In this case, the function is checking for rooks (both
    lowercase 'r' and uppercase 'R') on the board. The default value for the `, defaults to rook
    (optional)
    """
    chain = 1
    check = True
    for k, v in current_board.items():
        check = True
        if v == "r":
            for offset in rook_offsets:
                check = True
                chain = 1
                while check:
                    if (k[0] + chain * offset[0], k[1] + chain * offset[1]) in current_board and (current_board[(k[0] + chain * offset[0], k[1] + chain * offset[1])].isupper() or current_board[(k[0] + \
                            chain * offset[0], k[1] + chain * offset[1])] == ""):
                        
                        chain += 1
                        
                        if check:
                            moves_list['w'].append((k, (k[0] + (chain * offset[0] - offset[0]), k[1] + (chain * offset[1] - offset[1]))))
                            
                        if (k[0] + chain * offset[0] - offset[0], k[1] + chain * offset[1] - offset[1]) in current_board and current_board[(k[0] + chain * offset[0] - offset[0], k[1] + chain * offset[1] - offset[1])].isupper():
                            
                            check = False
                            chain += 1
                            moves_list['w'].append((k, (k[0] + (chain * offset[0] - offset[0] - offset[0]), k[1] + (chain * offset[1] - offset[1] - offset[1]))))
                    else:
                        check = False
        chain = 1
        check = True
        if v == "R":
            for offset in rook_offsets:
                check = True
                chain = 1
                while check:
                    if (k[0] + chain * offset[0], k[1] + chain * offset[1]) in current_board and (current_board[(k[0] + chain * offset[0], k[1] + chain * offset[1])].islower() or current_board[(k[0] + \
                            chain * offset[0], k[1] + chain * offset[1])] == ""):
                        
                        chain += 1
                        
                        if check:
                            moves_list['b'].append((k, (k[0] + (chain * offset[0] - offset[0]), k[1] + (chain * offset[1] - offset[1]))))
                            
                        if (k[0] + chain * offset[0] - offset[0], k[1] + chain * offset[1] - offset[1]) in current_board and current_board[(k[0] + chain * offset[0] - offset[0], k[1] + chain * offset[1] - offset[1])].islower():
                            
                            check = False
                            chain += 1
                            moves_list['b'].append((k, (k[0] + (chain * offset[0] - offset[0] - offset[0]), k[1] + (chain * offset[1] - offset[1] - offset[1]))))
                    else:
                        check = False
                        

File: ChessEngine/chessEngineCPP/rook.cpp
#include "rook.h"
#include <map>
#include <vector>

// Rook move offsets
const std::pair<int, int> rook_offsets[4] = {
    {1, 0}, {-1, 0}, {0, 1}, {0, -1}
};

void check_rook(const std::map<std::pair<int, int>, char>& current_board, 
                std::map<char, std::vector<std::pair<std::pair<int, int>, std::pair<int, int>>>>& moves_list) 
{
    for (const auto& piece : current_board) {
        std::pair<int, int> position = piece.first;
        char piece_type = piece.second;
        int chain = 1;
        
        // Check for white and black bishops
        if (piece_type == 'R' || piece_type == 'r') {
            for (const auto& offset : rook_offsets) {
                chain = 1;
                bool check = true;
                
                while (check) {
                    std::pair<int, int> target_position = {
                        position.first + chain * offset.first,
                        position.second + chain * offset.second
                    };
                    
                    // If target position is on the board
                    if (current_board.find(target_position) != current_board.end()) {
                        char target_piece = current_board.at(target_position);
                        if ((piece_type == 'R' && isupper(target_piece)) ||
                            (piece_type == 'r' && islower(target_piece))) {
                            check = false;
                            break;
                        }
                        // Determine if it's a valid move for black or white
                        if (piece_type == 'R' && (target_piece == ' ' || islower(target_piece))) {
                            moves_list['b'].push_back({position, target_position});
                        } else if (piece_type == 'r' && (target_piece == ' ' || isupper(target_piece))) {
                            moves_list['w'].push_back({position, target_position});
                        }
                        
                        // Stop if there's an enemy piece to capture
                        if ((piece_type == 'R' && islower(target_piece)) ||
                            (piece_type == 'r' && isupper(target_piece))) {
                            check = false;
                        }
                        
                    chain++;
                    } else {
                        check = false;
                    }
                }
            }
        }
    }
}
